aws/resource_aws_budgets_budget_test.go

```diff
@@ -0,0 +1,408 @@
+			return err
+		}
+
+		if err := checkBudgetTimePeriod(config, *b.Budget.TimePeriod); err != nil {
+			return err
+		}
+
+		if v, ok := b.Budget.CostFilters[config.FilterKey]; !ok || *v[len(v)-1] != config.FilterValue {
+			return fmt.Errorf("cost filter not set properly: %v", b.Budget.CostFilters)
+		}
+
+		return nil
+	}
+}
+
+func checkBudgetTimePeriod(config budgetTestConfig, timePeriod budgets.TimePeriod) error {
+	if end, _ := time.Parse("2006-01-02_15:04", config.TimePeriodEnd); *timePeriod.End != end {
+		return fmt.Errorf("TimePeriodEnd not set properly '%v' should be '%v'", *timePeriod.End, end)
+	}
+
+	if start, _ := time.Parse("2006-01-02_15:04", config.TimePeriodStart); *timePeriod.Start != start {
+		return fmt.Errorf("TimePeriodStart not set properly '%v' should be '%v'", *timePeriod.Start, start)
+	}
+
+	return nil
+}
+
+func checkBudgetCostTypes(config budgetTestConfig, costTypes budgets.CostTypes) error {
+	if strconv.FormatBool(*costTypes.IncludeCredit) != config.IncludeCredit {
+		return fmt.Errorf("IncludeCredit not set properly '%v' should be '%v'", *costTypes.IncludeCredit, config.IncludeCredit)
+	}
+
+	if strconv.FormatBool(*costTypes.IncludeOtherSubscription) != config.IncludeOtherSubscription {
+		return fmt.Errorf("IncludeOtherSubscription not set properly '%v' should be '%v'", *costTypes.IncludeOtherSubscription, config.IncludeOtherSubscription)
+	}
+
+	if strconv.FormatBool(*costTypes.IncludeRecurring) != config.IncludeRecurring {
+		return fmt.Errorf("IncludeRecurring not set properly  '%v' should be '%v'", *costTypes.IncludeRecurring, config.IncludeRecurring)
+	}
+
+	if strconv.FormatBool(*costTypes.IncludeRefund) != config.IncludeRefund {
+		return fmt.Errorf("IncludeRefund not set properly '%v' should be '%v'", *costTypes.IncludeRefund, config.IncludeRefund)
+	}
+
+	if strconv.FormatBool(*costTypes.IncludeSubscription) != config.IncludeSubscription {
+		return fmt.Errorf("IncludeSubscription not set properly '%v' should be '%v'", *costTypes.IncludeSubscription, config.IncludeSubscription)
+	}
+
+	if strconv.FormatBool(*costTypes.IncludeSupport) != config.IncludeSupport {
+		return fmt.Errorf("IncludeSupport not set properly '%v' should be '%v'", *costTypes.IncludeSupport, config.IncludeSupport)
+	}
+
+	if strconv.FormatBool(*costTypes.IncludeTax) != config.IncludeTax {
+		return fmt.Errorf("IncludeTax not set properly '%v' should be '%v'", *costTypes.IncludeTax, config.IncludeTax)
+	}
+
+	if strconv.FormatBool(*costTypes.IncludeUpfront) != config.IncludeUpfront {
+		return fmt.Errorf("IncludeUpfront not set properly '%v' should be '%v'", *costTypes.IncludeUpfront, config.IncludeUpfront)
+	}
+
+	if strconv.FormatBool(*costTypes.UseBlended) != config.UseBlended {
+		return fmt.Errorf("UseBlended not set properly '%v' should be '%v'", *costTypes.UseBlended, config.UseBlended)
+	}
+
+	return nil
+}
+
+func testCheckBudgetDestroy(budgetName string, provider *schema.Provider) error {
```

Comment:
This should be checking all possible resources in the acceptance testing Terraform state:

```go
func testAccCheckAWSBudgetsBudgetDestroy(s *terraform.State) error {
  conn := testAccProvider.Meta().(*AWSClient).budgetconn
  for _, rs := range s.RootModule().Resources {
    if rs.Type != "aws_budgets_budget" {
      continue
    }
    // rest of logic using rs.Primary.ID instead of budgetName
  }
  
  return nil
}
```